### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [311]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
csv_path = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
file_df = pd.read_csv(csv_path)


## Player Count

* Display the total number of players


In [312]:
# Total Number of Players (some players buy more than one item therefore I have to use value_counts)

formatted_file_df = file_df.drop_duplicates(subset ="SN") 
total_number_of_players = formatted_file_df['SN'].count()

total_players = {'Total Number of Players':[total_number_of_players]}
total_players_df = pd.DataFrame(total_players)
total_players_df


Total Number of Players
0                      576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [313]:
# Total Number of Unique Items

number_of_unique_items = file_df["Item Name"].value_counts()
number_of_unique_items_df = pd.DataFrame(number_of_unique_items)
number_of_unique_items = number_of_unique_items_df["Item Name"].count()

# Average Purchase Price

average_purchase_price = round(file_df["Price"].mean(), 2)

formatted_average_purchase_price = '${0:,.2f}'.format(average_purchase_price)

# Number of Purchases

total_number_of_purchases = file_df["Price"].count()

# Total Revenue

total_revenue = file_df["Price"].sum()

formatted_total_revenue = '${0:,.2f}'.format(total_revenue)


In [314]:
# Purchasing Analysis (Total)

purchasing_data = {'Number of Unique Items':[number_of_unique_items], 
                   'Average Price':[formatted_average_purchase_price], 
                   'Number of Purchases':[total_number_of_purchases], 
                   'Total Revenue': [formatted_total_revenue]
                  }

purchasing_analysis_total_df = pd.DataFrame(purchasing_data)
purchasing_analysis_total_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [315]:
# Count and Percentage of Male Players

male_players_df = formatted_file_df.loc[formatted_file_df["Gender"] == "Male"]
male_players_count = male_players_df["Gender"].count()

male_players_percentage = (male_players_count / formatted_file_df["Gender"].count()) * 100
formatted_male_players_percentage = '{0:,.2f}%'.format(male_players_percentage)


# Count and Percentage of Female Players

female_players_df = formatted_file_df.loc[formatted_file_df["Gender"] == "Female"]
female_players_count = female_players_df["Gender"].count()

female_players_percentage = (female_players_count / formatted_file_df["Gender"].count()) * 100
formatted_female_players_percentage = '{0:,.2f}%'.format(female_players_percentage)


# Count and Percentage of Other / Non-Disclosed Players

other_players_df = formatted_file_df.loc[formatted_file_df["Gender"] == "Other / Non-Disclosed"]
other_players_count = other_players_df["Gender"].count()

other_players_percentage = (other_players_count / formatted_file_df["Gender"].count()) * 100
formatted_other_players_percentage = '{0:,.2f}%'.format(other_players_percentage)


# Gender Demographics

gender_demographics = {'Total Count':[male_players_count, female_players_count, other_players_count], 
                       'Percentage of Players': [formatted_male_players_percentage, formatted_female_players_percentage, formatted_other_players_percentage]
                      }


gender_demographics_df = pd.DataFrame(gender_demographics)

gender_demographics_df_new = gender_demographics_df.rename(index={0: 'Male', 1: "Female", 2:"Other / Non-Disclosed"})
gender_demographics_df_new



Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [316]:
# Use groupby based on genders

gender_group = file_df.groupby(["Gender"])
new_gender_group = gender_group.count()

# Delete unnecessary rows
del new_gender_group["SN"]
del new_gender_group["Age"]
del new_gender_group["Item ID"]
del new_gender_group["Item Name"]
del new_gender_group["Price"]

male_purchase_count_df = file_df.loc[file_df["Gender"] == "Male"]
female_purchase_count_df = file_df.loc[file_df["Gender"] == "Female"]
other_purchase_count_df = file_df.loc[file_df["Gender"] == "Other / Non-Disclosed"]

male_purchase_count = file_df.loc[file_df["Gender"] == "Male"]["Gender"].count()
female_purchase_count = file_df.loc[file_df["Gender"] == "Female"]["Gender"].count()
other_purchase_count = file_df.loc[file_df["Gender"] == "Other / Non-Disclosed"]["Gender"].count()

# Average Purchase Price

male_average_purchase_price = male_purchase_count_df['Price'].sum() / male_purchase_count

female_average_purchase_price = female_purchase_count_df['Price'].sum() / female_purchase_count

other_average_purchase_price = other_purchase_count_df['Price'].sum() / other_purchase_count


# Total Purchase Value

male_total_purchase_value = male_purchase_count_df['Price'].sum()

female_total_purchase_value = female_purchase_count_df['Price'].sum()

other_total_purchase_value = other_purchase_count_df['Price'].sum()


# Avg Total Purchase per Person

avg_total_purchase_per_male = male_total_purchase_value / male_players_count

avg_total_purchase_per_female = female_total_purchase_value / female_players_count

avg_total_purchase_per_other = other_total_purchase_value / other_players_count


# Adding new columns to the dataframe

av_pu_price = [female_average_purchase_price, male_average_purchase_price, other_average_purchase_price]
to_pu_value = [female_total_purchase_value, male_total_purchase_value, other_total_purchase_value]
av_to_pu_per_person = [avg_total_purchase_per_female, avg_total_purchase_per_male, avg_total_purchase_per_other]

new_gender_group.loc[:, "Average Purchase Price"] = av_pu_price
new_gender_group.loc[:, "Total Purchase Value"] = to_pu_value
new_gender_group.loc[:, "Avg Total Purchase per Person"] = av_to_pu_per_person


# Formatting
new_gender_group["Average Purchase Price"] = new_gender_group["Average Purchase Price"].astype(float).map("${:.2f}".format)
new_gender_group["Total Purchase Value"] = new_gender_group["Total Purchase Value"].astype(float).map("${:,.2f}".format)
new_gender_group["Avg Total Purchase per Person"] = new_gender_group["Avg Total Purchase per Person"].astype(float).map("${:.2f}".format)

final_gender_group = new_gender_group.rename(columns= {"Purchase ID": "Purchase Count"})

final_gender_group

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [317]:
# Establish bins for ages

bins = [0, 9, 14, 19, 24, 29, 34, 39, 200]

group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34",
                "35-39", "40+"]

# Categorize the existing players using the age bins.

age_group_formatted_file_df = formatted_file_df.copy()
age_group_formatted_file_df["Age Group"] = pd.cut(formatted_file_df["Age"], bins, labels=group_names)


# Calculate the numbers and percentages by age group

less_than_10 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '<10']["SN"].count()
percentage_less_than_10 = less_than_10 / total_number_of_players * 100


from_10_to_14 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '10-14']["SN"].count()
percentage_from_10_to_14 = from_10_to_14 / total_number_of_players * 100


from_15_to_19 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '15-19']["SN"].count()
percentage_from_15_to_19 = from_15_to_19 / total_number_of_players * 100


from_20_to_24 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '20-24']["SN"].count()
percentage_from_20_to_24 = from_20_to_24 / total_number_of_players * 100


from_25_to_29 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '25-29']["SN"].count()
percentage_from_25_to_29 = from_25_to_29 / total_number_of_players * 100


from_30_to_34 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '30-34']["SN"].count()
percentage_from_30_to_34 = from_30_to_34 / total_number_of_players * 100


from_35_to_39 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '35-39']["SN"].count()
percentage_from_35_to_39 = from_35_to_39 / total_number_of_players * 100


greater_than_40 = age_group_formatted_file_df.loc[age_group_formatted_file_df["Age Group"] == '40+']["SN"].count()
percentage_greater_than_40 = greater_than_40 / total_number_of_players * 100

# Create a summary data frame to hold the results

age_demographics_data = {'Total Count':[less_than_10, from_10_to_14, from_15_to_19, from_20_to_24, from_25_to_29, from_30_to_34, from_35_to_39, greater_than_40], 
                       'Percentage of Players': [percentage_less_than_10, percentage_from_10_to_14, percentage_from_15_to_19, percentage_from_20_to_24, percentage_from_25_to_29, percentage_from_30_to_34, percentage_from_35_to_39, percentage_greater_than_40]
                      }

age_demographics_df = pd.DataFrame(age_demographics_data)
age_demographics_df["Percentage of Players"] = age_demographics_df["Percentage of Players"].astype(float).map('{0:,.2f}%'.format)
age_demographics_df

age_demographics_df_new = age_demographics_df.rename(index={0: "<10", 1: '10-14', 2: "15-19", 3:"20-24", 4:"25-29", 5:"30-34", 6:"35-39", 7:"40+"})
age_demographics_df_new




Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [318]:
# Establish bins for ages

bins = [0, 9, 14, 19, 24, 29, 34, 39, 200]

group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34",
                "35-39", "40+"]

# Categorize the existing players using the age bins.

age_group_file_df = file_df.copy()
age_group_file_df["Age Ranges"] = pd.cut(file_df["Age"], bins, labels=group_names)

# Use groupby based on age ranges

new_age_group_file_df = age_group_file_df.groupby(["Age Ranges"])

# Delete unnecessary rows

age_group_count = new_age_group_file_df.count()

del age_group_count["SN"]
del age_group_count["Age"]
del age_group_count["Gender"]
del age_group_count["Item ID"]
del age_group_count["Item Name"]
del age_group_count["Price"]



# Run basic calculations to obtain avg. purchase price, avg. purchase total per person, total value

less_than_10_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "<10"]["Price"].mean()
less_than_10_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "<10"]["Price"].sum()
less_than_10_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "<10"]["Price"].sum() / less_than_10


from_10_to_14_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "10-14"]["Price"].mean()
from_10_to_14_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "10-14"]["Price"].sum()
from_10_to_14_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "10-14"]["Price"].sum() / from_10_to_14


from_15_to_19_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "15-19"]["Price"].mean()
from_15_to_19_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "15-19"]["Price"].sum()
from_15_to_19_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "15-19"]["Price"].sum() / from_15_to_19


from_20_to_24_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "20-24"]["Price"].mean()
from_20_to_24_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "20-24"]["Price"].sum()
from_20_to_24_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "20-24"]["Price"].sum() / from_20_to_24


from_25_to_29_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "25-29"]["Price"].mean()
from_25_to_29_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "25-29"]["Price"].sum()
from_25_to_29_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "25-29"]["Price"].sum() / from_25_to_29


from_30_to_34_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "30-34"]["Price"].mean()
from_30_to_34_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "30-34"]["Price"].sum()
from_30_to_34_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "30-34"]["Price"].sum() / from_30_to_34


from_35_to_39_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "35-39"]["Price"].mean()
from_35_to_39_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "35-39"]["Price"].sum()
from_35_to_39_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "35-39"]["Price"].sum() / from_35_to_39


greater_than_40_avg_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "40+"]["Price"].mean()
greater_than_40_total_pur_price = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "40+"]["Price"].sum()
greater_than_40_avg_pur_price_per_person = age_group_file_df.loc[age_group_file_df["Age Ranges"] == "40+"]["Price"].sum() / greater_than_40


# Adding new columns to the dataframe

a_p_price = [less_than_10_avg_pur_price, from_10_to_14_avg_pur_price, from_15_to_19_avg_pur_price, from_20_to_24_avg_pur_price, from_25_to_29_avg_pur_price, from_30_to_34_avg_pur_price, from_35_to_39_avg_pur_price, greater_than_40_avg_pur_price]
t_p_value = [less_than_10_total_pur_price, from_10_to_14_total_pur_price, from_15_to_19_total_pur_price, from_20_to_24_total_pur_price, from_25_to_29_total_pur_price, from_30_to_34_total_pur_price, from_35_to_39_total_pur_price, greater_than_40_total_pur_price]
a_t_p_per_person = [less_than_10_avg_pur_price_per_person, from_10_to_14_avg_pur_price_per_person, from_15_to_19_avg_pur_price_per_person, from_20_to_24_avg_pur_price_per_person, from_25_to_29_avg_pur_price_per_person, from_30_to_34_avg_pur_price_per_person, from_35_to_39_avg_pur_price_per_person, greater_than_40_avg_pur_price_per_person]

age_group_count.loc[:, "Average Purchase Price"] = a_p_price
age_group_count.loc[:, "Total Purchase Value"] = t_p_value
age_group_count.loc[:, "Avg Total Purchase per Person"] = a_t_p_per_person


# Formatting
age_group_count["Average Purchase Price"] = age_group_count["Average Purchase Price"].astype(float).map("${:.2f}".format)
age_group_count["Total Purchase Value"] = age_group_count["Total Purchase Value"].astype(float).map("${:,.2f}".format)
age_group_count["Avg Total Purchase per Person"] = age_group_count["Avg Total Purchase per Person"].astype(float).map("${:.2f}".format)

age_group_count_renamed = age_group_count.rename(columns = {"Purchase ID" : "Purchase Count"})

age_group_count_renamed


Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [324]:
# Grouping the main dataframe(file_df) by "SN" column, and converting it into a new dataframe

SN_group = file_df.groupby(["SN"])
groupby_df = SN_group.sum()

# Deleting unnecessary columns

del groupby_df["Purchase ID"]
del groupby_df["Item ID"]
del groupby_df["Age"]

# Calculating Required Values

purc_id_counts = SN_group["Purchase ID"].count().values

groupby_df.loc[:, "Purchase Count"] = purc_id_counts

groupby_df.loc[:, "Average Purchase Price"] = round(groupby_df["Price"] / groupby_df["Purchase Count"], 2)

# Sorting the dataframe

top_spenders = groupby_df.sort_values(["Price"], ascending=False)

# Renaming Price column

top_spenders = top_spenders.rename(columns={"Price": "Total Purchase Value"})

# Organizing the dataframe

top_spenders_organized = top_spenders[["Purchase Count", "Average Purchase Price", "Total Purchase Value"]]

top_spenders_organized.head(5)

# Formatting the values in the columns

top_spenders_organized["Average Purchase Price"] = top_spenders_organized["Average Purchase Price"].astype(float).map("${:.2f}".format)
top_spenders_organized["Total Purchase Value"] = top_spenders_organized["Total Purchase Value"].astype(float).map("${:.2f}".format)

top_spenders_organized.head()
    

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [326]:
# Grouping the main dataframe(file_df) by "Item ID" and "Item Name" columns, and converting it into a new dataframe

most_popular = file_df.groupby(["Item ID", "Item Name"]).count()

# Deleting unnecessary columns

del most_popular["Purchase ID"]
del most_popular["SN"]
del most_popular["Age"]
del most_popular["Gender"]

# Calculating Required Values and organizing the columns

mp_item_count = file_df.groupby(["Item ID"])["Purchase ID"].count().values
mp_item_sum = file_df.groupby(["Item ID"])["Price"].sum().values

most_popular_renamed = most_popular.rename(columns={"Price": "Purchase Count"})

most_popular_renamed.loc[:, "Total Purchase Value"] = mp_item_sum

most_popular_renamed.loc[:, "Average Purchase Price"] = round(most_popular_renamed["Total Purchase Value"] / most_popular_renamed["Purchase Count"], 2)

most_popular_renamed = most_popular_renamed.rename(columns={"Average Purchase Price": "Item Price"})

most_popular_items = most_popular_renamed[["Purchase Count", "Item Price", "Total Purchase Value"]]

# Sorting the Dataframe

formatted_most_popular_items = most_popular_items.sort_values(["Purchase Count"], ascending=False)


# Formatting the values in the columns

formatted_most_popular_items["Item Price"] = formatted_most_popular_items["Item Price"].astype(float).map("${:.2f}".format)
formatted_most_popular_items["Total Purchase Value"] = formatted_most_popular_items["Total Purchase Value"].astype(float).map("${:.2f}".format)

formatted_most_popular_items.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Cells 

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [327]:
# Sorting the above table by total purchase value in descending order

most_profitable_cells = most_popular_items.sort_values(["Total Purchase Value"], ascending=False)

# Formatting the values in the columns

most_profitable_cells["Item Price"] = most_profitable_cells["Item Price"].astype(float).map("${:.2f}".format)
most_profitable_cells["Total Purchase Value"] = most_profitable_cells["Total Purchase Value"].astype(float).map("${:.2f}".format)

most_profitable_cells.head()

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# Observations Based on Data

#### 1. Age Ranges 15-19 and 20-24 combined contributes to  by far more than half of the Total Purchase Value (total dollars made by the company by selling optional items).

#### 2. Age Group 35-39 spend the most money per person but the number of players pertaining to that age group is cosiderably low.

#### 3. Male players constitute more than 80% of all the players and their spending (approximately 82% of the total is aligned with their percentage.